pkg/segment/query/processor/headcommand.go

```diff
@@ -30,11 +32,83 @@ type headProcessor struct {
 	numRecordsSent uint64
 }
 
+func (p *headProcessor) processHeadExpr(iqr *iqr.IQR) (*iqr.IQR, error) {
+	if iqr == nil || p.options.Done {
+		return nil, io.EOF
+	}
+
+	requiredFields := p.options.BoolExpr.GetFields()
+	records, err := iqr.ReadColumns(requiredFields)
+	if err != nil {
+		return nil, toputils.TeeErrorf("headProcessor.processHeadExpr: failed to read columns, requiredFields: %v, err: %v", requiredFields, err)
+	}
+
+	row := 0
+	for row < iqr.NumberOfRecords() && !p.options.Done && p.numRecordsSent < p.options.MaxRows {
+		fieldToValue := make(map[string]utils.CValueEnclosure, len(requiredFields))
+		for _, field := range requiredFields {
+			fieldToValue[field] = records[field][row]
+		}
+
+		conditionPassed, err := p.options.BoolExpr.Evaluate(fieldToValue)
+		if err != nil {
+			nullFields, errGetNullFields := p.options.BoolExpr.GetNullFields(fieldToValue)
+			if errGetNullFields != nil {
+				return nil, toputils.TeeErrorf("headProcessor.processHeadExpr: error while getting null fields, err: %v", errGetNullFields)
+			} else if len(nullFields) > 0 {
+				// evaluation failed due to null fields
+				if p.options.Null {
+					// include the records since null is allowed
+				} else if p.options.Keeplast {
+					p.options.Done = true
+					// if keeplast is true, we need to include this record
+				} else {
+					// if null is not allowed and keeplast is false, we are done processing
+					p.options.Done = true
+					break
+				}
+			} else {
+				return nil, toputils.TeeErrorf("headProcessor.processHeadExpr: Error while evaluating boolean expr, err: %v", err)
+			}
+		} else {
+			if !conditionPassed {
+				// false condition so adding last record if keeplast
+				if p.options.Keeplast {
+					p.options.Done = true
+				} else {
+					p.options.Done = true
+					break
+				}
+			}
+		}
+		row++
+		p.numRecordsSent++
+	}
+
+	if p.numRecordsSent == p.options.MaxRows {
+		p.options.Done = true
+	}
```

Comment:
Shouldn't this `if` be inside the loop?

Reply:
No we can just check this after exit, because the loop invariant has the condition `p.numRecordsSent < p.options.MaxRows`